zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,992 @@
+        }
+
+        self.payments.sort_by(Payment::compare_normalized(params));
+    }
+
+    #[cfg(all(test, feature = "test-dependencies"))]
+    pub(in crate::zip321) fn normalize_and_eq<P: consensus::Parameters>(
+        params: &P,
+        a: &mut TransactionRequest,
+        b: &mut TransactionRequest,
+    ) -> bool {
+        a.normalize(params);
+        b.normalize(params);
+
+        a == b
+    }
+
+    /// Convert this request to a URI string.
+    ///
+    /// Returns None if the payment request is empty.
+    pub fn to_uri<P: consensus::Parameters>(&self, params: &P) -> Option<String> {
+        fn push_params(
+            payment: &Payment,
+            payment_index: Option<usize>,
+            query_params: &mut Vec<String>,
+        ) {
+            if let Some(amount_p) = render::amount_param(payment.amount, payment_index) {
+                query_params.push(amount_p);
+            }
+
+            if let Some(memo_p) = payment
+                .memo
+                .as_ref()
+                .map(|m| render::memo_param(&m, payment_index))
+            {
+                query_params.push(memo_p);
+            }
+
+            if let Some(label_p) = payment
+                .label
+                .as_ref()
+                .map(|m| render::str_param("label", &m, payment_index))
+            {
+                query_params.push(label_p);
+            }
+
+            if let Some(message_p) = payment
+                .message
+                .as_ref()
+                .map(|m| render::str_param("message", &m, payment_index))
+            {
+                query_params.push(message_p);
+            }
+
+            for (name, value) in &payment.other_params {
+                query_params.push(render::str_param(&name, &value, payment_index))
+            }
+        }
+
+        let mut query_params = vec![];
+        if self.payments.len() == 1 {
+            self.payments.get(0).map(|payment| {
+                push_params(payment, None, &mut query_params);
+                format!(
+                    "zcash:{}?{}",
+                    payment.recipient_address.encode(params),
+                    query_params.join("&")
+                )
+            })
+        } else {
+            for (i, payment) in (&self.payments).into_iter().enumerate() {
+                query_params.push(render::addr_param(
+                    params,
+                    &payment.recipient_address,
+                    Some(i),
+                ));
+                push_params(payment, Some(i), &mut query_params);
+            }
+
+            Some(format!("zcash:?{}", query_params.join("&")))
+        }
+    }
+
+    /// Parse the provided URI to a payment request value.
+    pub fn from_uri<P: consensus::Parameters>(params: &P, uri: &str) -> Result<Self, String> {
+        // Parse the leading zcash:<address>
+        let (rest, primary_addr_param) =
+            parse::lead_addr(&params.clone())(uri).map_err(|e| e.to_string())?;
```

Comment:
Oh... this was a hangover from a time when `lead_addr` returned a closure. I gave up on trying to write this as all combinators because the borrow checker gave me such fits.